File: Blind/63.cpp
#include <bits/stdc++.h>
using namespace std;

class MedianFinder {
public:
	//内側がtopになるように
	priority_queue<int> left;
	priority_queue<int, vector<int>, greater<int>> right;

	MedianFinder() {}

	void addNum(int num) {
		if (left.empty())
			left.push(num);
		else if (num <= left.top())
			left.push(num);
		else
			right.push(num);
		//leftの方が１つ多いか同じ大きさにする
		if (left.size() - 1 > right.size())
		{
			int temp = left.top();
			left.pop();
			right.push(temp);
		}
		else if (left.size() < right.size())
		{
			int temp = right.top();
			right.pop();
			left.push(temp);
		}
	}

	double findMedian() {
		if (left.size() == right.size())
			return (left.top() + right.top()) / 2.0;
		else
			return left.top();
	}
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

class MedianFinder {
public:
	priority_queue<int> left;
	priority_queue<int, vector<int>, greater<int>> right;

	MedianFinder() {}

	void addNum(int num) {
		if (left.empty())
			left.push(num);
		else if (num <= left.top())
			left.push(num);
		else
			right.push(num);
		if (left.size() - 1 > right.size())
		{
			int temp = left.top();
			left.pop();
			right.push(temp);
		}
		else if (left.size() < right.size())
		{
			int temp = right.top();
			right.pop();
			left.push(temp);
		}
		priority_queue<int> left_copy(left);
		priority_queue<int, vector<int>, greater<int>> right_copy(right);
		cout << "left is ";
		while(!left_copy.empty())
		{
			cout << left_copy.top() << " ";
			left_copy.pop();
		}
		cout << " right is ";
		while(!right_copy.empty())
		{
			cout << right_copy.top() << " ";
			right_copy.pop();
		}
		cout << endl;
	}

	double findMedian() {
		if (left.size() == right.size())
			return (left.top() + right.top()) / 2.0;
		else
			return left.top();
	}
};

int main() {
	MedianFinder* obj = new MedianFinder();
	obj->addNum(1);
	cout << "ans is " << obj->findMedian() << endl;
	obj->addNum(2);
	cout << "ans is " << obj->findMedian() << endl;
	obj->addNum(0);
	cout << "ans is " << obj->findMedian() << endl;
	obj->addNum(10);
	cout << "ans is " << obj->findMedian() << endl;
	obj->addNum(9);
	cout << "ans is " << obj->findMedian() << endl;
	obj->addNum(8);
	cout << "ans is " << obj->findMedian() << endl;
	obj->addNum(7);
	cout << "ans is " << obj->findMedian() << endl;
	obj->addNum(-1);
	cout << "ans is " << obj->findMedian() << endl;
	obj->addNum(11);
	cout << "ans is " << obj->findMedian() << endl;
	return 0;
}

File: Blind/63.go
package main

import "fmt"

// 関数内で
// p := struct {
// 	name string
// 	age int
// }{
// 	name: "Gopher", 
// 	age: 10,
// }

type MedianFinder struct {
	data []float64
}

func Constructor() MedianFinder {
	return MedianFinder{
		data: make([]float64, 0, 1000),
	}
}

func (this *MedianFinder) AddNum(num int) {
	this.data = append(this.data, float64(num))
}

func (this *MedianFinder) FindMedian() float64 {
	d_len := len(this.data)
	if d_len&1 == 1 {
		return this.data[d_len/2]
	} else {
		return (this.data[d_len/2-1] + this.data[d_len/2]) / 2
	}
}

func main() {
	obj := Constructor()
	obj.AddNum(6)
	fmt.Println(obj.FindMedian())
	obj.AddNum(10)
	fmt.Println(obj.FindMedian())
	obj.AddNum(2)
	fmt.Println(obj.FindMedian())
	obj.AddNum(6)
	fmt.Println(obj.FindMedian())
	obj.AddNum(5)
	fmt.Println(obj.FindMedian())
	obj.AddNum(0)
	fmt.Println(obj.FindMedian())
	obj.AddNum(6)
	fmt.Println(obj.FindMedian())
	obj.AddNum(3)
	fmt.Println(obj.FindMedian())
	obj.AddNum(1)
	fmt.Println(obj.FindMedian())
	obj.AddNum(0)
	fmt.Println(obj.FindMedian())
	obj.AddNum(0)
	fmt.Println(obj.FindMedian())
}


/**
 * Your MedianFinder object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddNum(num);
 * param_2 := obj.FindMedian();
 */


//  ["MedianFinder","addNum","findMedian","addNum","findMedian"]
//  [[],[6],[],[10],[]]
//  に対応するmain関数は，
//  func main() {
// 	 obj := Constructor()
// 	 obj.AddNum(6)
// 	 fmt.Println(obj.FindMedian())
// 	 obj.AddNum(10)
// 	 fmt.Println(obj.FindMedian())
//  }
//  です。
//  ["MedianFinder","addNum","findMedian","addNum","findMedian","addNum","findMedian","addNum","findMedian","addNum","findMedian","addNum","findMedian","addNum","findMedian","addNum","findMedian","addNum","findMedian","addNum","findMedian","addNum","findMedian"]
//  [[],[6],[],[10],[],[2],[],[6],[],[5],[],[0],[],[6],[],[3],[],[1],[],[0],[],[0],[]]
//  に対応するmain関数を書いてください。